src/Chainweb/Mempool/InMem.hs

```diff
@@ -201,60 +206,106 @@ markValidatedInMem :: MVar (InMemoryMempoolData t)
                    -> IO ()
 markValidatedInMem lock txs = withMVarMasked lock $ \mdata -> do
     let pref = _inmemPending mdata
-    modifyIORef' pref $ \psq -> foldl' (flip HashMap.delete) psq txs
+    modifyIORef' pref $ \psq -> foldl' (flip HM.delete) psq txs
 
 maxNumPending :: Int
 maxNumPending = 10000
 
 ------------------------------------------------------------------------------
+
+-- | Validation: A short-circuiting variant of this check that fails outright at
+-- the first detection of any validation failure on any Transaction.
+--
 insertCheckInMem
-    :: InMemConfig t    -- ^ in-memory config
+    :: forall t
+    .  InMemConfig t    -- ^ in-memory config
     -> MVar (InMemoryMempoolData t)  -- ^ in-memory state
     -> Vector t  -- ^ new transactions
-    -> IO (Vector (TransactionHash, Maybe InsertError))
+    -> IO (Either (TransactionHash, InsertError) ())
 insertCheckInMem cfg lock txs = do
     now <- getCurrentTimeIntegral
     badmap <- withMVarMasked lock $ readIORef . _inmemBadMap
-    let hashes = V.map hasher txs
-    let txhashes = V.zip hashes txs
-    let checks = [ (InsertErrorOversized, sizeOK)
-                 , (InsertErrorInvalidTime, ttlCheck now)
-                 , (InsertErrorBadlisted, notInBadMap badmap)
-                 ]
-    let out1 = V.map (\tx -> runPreChecks checks tx) txhashes
-    out2 <- _inmemPreInsertCheck cfg txs
-    return $! V.zip hashes (V.zipWith (<|>) out1 out2)
 
+    -- We hash the tx here and pass it around around to avoid needing to repeat
+    -- the hashing effort.
+    let withHashes :: Either (TransactionHash, InsertError) (Vector (TransactionHash, t))
+        withHashes = for txs $ \tx ->
+          let !h = hasher tx
+          in bimap (h,) (h,) $ validateOne cfg badmap now tx h
+
+    case withHashes of
+        Left _ -> pure $ void withHashes
+        Right r -> void . sequenceA <$> _inmemPreInsertBatchChecks cfg r
```

Comment:
As mentioned below, the one-off checks and batch checks are now separated. Within `withHashes`, we fail utterly at the first sign of *any* TX not passing validation. If all pass, then we move on to the batch checks, which unfortunately have a semi-inefficient `sequenceA` (at least w.r.t. short-circuiting).